### PR TITLE
Enhancement intersection function 3265

### PR DESCRIPTION
**Pull request by ldelille**
This is a proposition to fix issue #3265 - Enhancement for the intersection of two graphs. 

The code does not raise a Value Error anymore when we try to intersect two graphs that don't have the same node sets. It now returns a graph which contains all the nodes that are the intersection of the two node sets and edges that are the intersection of the two edge sets.

I added a few tests and modified those which where expecting an error when running intersection function for two different node sets. I updated documentation accordingly.

### CODE REVIEW

File: networkx/algorithms/operators/binary.py [hunk 1]
```diff
@@ -137,15 +137,13 @@ def disjoint_union(G, H):
...
+    """Returns a new graph that contains only the nodes and the edges that exist in
     both G and H.
 
-    The node sets of H and G must be the same.
-
     Parameters
     ----------
     G,H : graph
-       A NetworkX graph.  G and H must have the same node sets.
+       A NetworkX graph. G and H can have different node sets but must be both graphs or both multigraphs.
```
Comment (dschult) [inline networkx/algorithms/operators/binary.py:146]: Could one be a MultiGraph and the other be a Graph?
What happens in that case?

Comment (ldelille) [reply networkx/algorithms/operators/binary.py:146]: It raises a NetworkX error, I didn't find any straightforward definition for the intersection of graphs of different types.

**Review by dschult — APPROVED** [review]
Looks good to me!   Thanks!

**Review by jarrodmillman — APPROVED** [review]

**Comment by stefanv** [comment]
What happens to edge attributes?  Should whatever happens be explicitly mentioned?

**Comment by ldelille** [comment]
> What happens to edge attributes? Should whatever happens be explicitly mentioned?

Like node and graph attributes, they are not copied to the new graph. It is mentioned in the Notes of the function (with a few lines of code to describe how to get the intersection graph with all attributes).

**Comment by stefanv** [comment]
Great, thanks for clarifying!

**Pull request merged**